bindings/java/src/main/com/apple/foundationdb/tuple/TupleUtil.java

```diff
@@ -159,318 +261,325 @@ static int getCodeFor(Object o) {
+			BigInteger res = new BigInteger(intBytes);
+			state.add(res, start + n + 1);
 		}
-		if(code == NEG_INT_START) {
+		else if(code == NEG_INT_START) {
 			int n = (rep[start] ^ 0xff) & 0xff;
-			BigInteger origValue = new BigInteger(ByteArrayUtil.join(new byte[]{0x00}, Arrays.copyOfRange(rep, start+1, start+n+1)));
+			byte[] intBytes = new byte[n + 1];
+			System.arraycopy(rep, start + 1, intBytes, 1, n);
+			BigInteger origValue = new BigInteger(intBytes);
 			BigInteger offset = BigInteger.ONE.shiftLeft(n*8).subtract(BigInteger.ONE);
-			return new DecodeResult(start + n + 1, origValue.subtract(offset));
+			state.add(origValue.subtract(offset), start + n + 1);
 		}
-		if(code > NEG_INT_START && code < POS_INT_END) {
+		else if(code > NEG_INT_START && code < POS_INT_END) {
 			// decode a long
-			byte[] longBytes = new byte[9];
-			boolean upper = code >= INT_ZERO_CODE;
-			int n = upper ? code - 20 : 20 - code;
+			boolean positive = code >= INT_ZERO_CODE;
+			int n = positive ? code - INT_ZERO_CODE : INT_ZERO_CODE - code;
 			int end = start + n;
 
-			if(rep.length < end) {
-				throw new RuntimeException("Invalid tuple (possible truncation)");
+			if(last < end) {
+				throw new IllegalArgumentException("Invalid tuple (possible truncation)");
 			}
 
-			System.arraycopy(rep, start, longBytes, longBytes.length-n, n);
-			if (!upper)
-				for(int i=longBytes.length-n; i<longBytes.length; i++)
-					longBytes[i] = (byte)(longBytes[i] ^ 0xff);
-
-			BigInteger val = new BigInteger(longBytes);
-			if (!upper) val = val.negate();
-
-			// Convert to long if in range -- otherwise, leave as BigInteger.
-			if (val.compareTo(BigInteger.valueOf(Long.MIN_VALUE))<0||
-				val.compareTo(BigInteger.valueOf(Long.MAX_VALUE))>0) {
-				// This can occur if the thing can be represented with 8 bytes but not
-				// the right sign information.
-				return new DecodeResult(end, val);
+			if(positive && (n < Long.BYTES || rep[start] > 0)) {
+				long res = 0L;
+				for(int i = start; i < end; i++) {
+					res = (res << 8) + (rep[i] & 0xff);
+				}
+				state.add(res, end);
+			}
+			else if(!positive && (n < Long.BYTES || rep[start] < 0)) {
+				long res = ~0L;
+				for(int i = start; i < end; i++) {
+					res = (res << 8) + (rep[i] & 0xff);
+				}
+				state.add(res + 1, end);
+			}
+			else {
+				byte[] longBytes = new byte[9];
+				System.arraycopy(rep, start, longBytes, longBytes.length-n, n);
+				if (!positive)
+					for(int i=longBytes.length-n; i<longBytes.length; i++)
+						longBytes[i] = (byte)(longBytes[i] ^ 0xff);
+
+				BigInteger val = new BigInteger(longBytes);
+				if (!positive) val = val.negate();
+
+				// Convert to long if in range -- otherwise, leave as BigInteger.
+				if (val.compareTo(LONG_MIN_VALUE) >= 0 && val.compareTo(LONG_MAX_VALUE) <= 0) {
+					state.add(val.longValue(), end);
+				} else {
+					// This can occur if the thing can be represented with 8 bytes but not
+					// the right sign information.
```

Comment:
What does "not the right sign information" mean?

Reply:
I reworded this comment to try and be more clear.